### ISSUE TITLE
FileHandles not released by Dispatcher

### ISSUE THREAD

**Issue by cdunn2001** [nim-lang/Nim#5725]
I've put a fairly simply test-case on the `never-finish` branch here:
* https://github.com/pb-cdunn/nim-multiproc/tree/never-finish

That should make this problem easy to replicate.

In `multiproc.nim`, if I move the `hasPendingOpertions()/poll()` loop ahead of the `finishParent()` loop, then it hangs. As it is, it croaks on an unregistered file-descriptor (Ubuntu14), or hangs (Macosx).

As an alternative, I would not mind creating a brand-new global Dispatcher, if that makes it easier to release the file-handles. I've filed an Issue (#5717) separately for that.

**Comment by bpr** [comment]
Using the latest devel, once I changed the type of pid to posix.Pid it compiled on Ubuntu 14.04 and behaved exactly as yours. OS X version required no such such change, same behavior.

If you're trying to make an analogue for Python's multiprocessing, I'd suggest leaving out the async calls and any dependence on asyncdispatch until everything else is working (more like the old module which didn't have apply_async), and only then add it in if you really need it. Sorry, I know that sounds unhelpful. My perception is that that module is not yet rock solid.

**Comment by cdunn2001** [comment]
@bpr, We [updated](https://github.com/pb-cdunn/nim-multiproc/pull/1) `Pid` for **Nim/devel**. Thanks for spotting that.

Yes, I could try using threads to communicate with the forks, using blocking pipes. But I still think there are problems in how asyncdispatch deals with file-descriptors.

**Comment by dom96** [comment]
Does this affect both the default and upcoming async dispatchers?

**Comment by pb-cdunn** [comment]
Slightly different crash.

I copied `lib/upcoming/asyncdispatcher.nim` over `lib/pure/` and re-built my program. It doesn't get as far:
```
waitFor result[0]:false
In prepareChild()
 -Closing fds:0x2b136204b088[3, 7, 5, 6, 8, 11]

Traceback (most recent call last)
main.nim(61)             main
main.nim(57)             main
main.nim(33)             example3
multiproc.nim(266)       newRpcPool
multiproc.nim(248)       newRpcFork
multiproc.nim(160)       runChild
asyncdispatch.nim(1189)  waitFor
asyncdispatch.nim(1202)  poll
ioselectors_epoll.nim(418) selectInto
system.nim(3533)         raiseAssert
system.nim(2643)         sysFatal
Error: unhandled exception:
  not (pkey.ident == 0)  [AssertionError]
...
closing Pool
parent sendZero()
parent recv zero
Traceback (most recent call last)
main.nim(61)             main
main.nim(57)             main
main.nim(45)             example3
multiproc.nim(278)       closePool
multiproc.nim(200)       sendZero
asyncdispatch.nim(1189)  waitFor
asyncdispatch.nim(1202)  poll
ioselectors_epoll.nim(418) selectInto
system.nim(3533)         raiseAssert
system.nim(2643)         sysFatal
Error: unhandled exception:
  not (pkey.ident == 0)  [AssertionError]

closing Pool
parent sendZero()
parent recv zero
Traceback (most recent call last)
main.nim(61)             main
main.nim(57)             main
main.nim(45)             example3
multiproc.nim(278)       closePool
multiproc.nim(200)       sendZero
asyncdispatch.nim(1189)  waitFor
asyncdispatch.nim(1198)  poll
Error: unhandled exception: No handles or timers registered in dispatcher. [ValueError]
```
(Note that there are two processes writing to stdout.)

That is using Nim@27b571dd9598d144a1d15963d9c73e75d7d764c4, the tip of `devel` this morning.

**Comment by cheatfate** [comment]
@pb-cdunn this looks like you doing something incorrectly, e.g. you have problems with inherited handles. Because with upcoming asyncdispatch assertions, i can say `epoll` returned file descriptor which was not registered with asyncdispatch/ioselectors module, so something wrong with your code...

**Comment by cdunn2001** [comment]
@cheatfate, I would love for that to be the case. But it's a very simple program. (And if you want, you can remove the msgpack part and communicate a long string of an agreed size.) The program tells me when it registers a FH, in both processes, and when it deregisters. I think something is being held by a library. Maybe I have subtly misunderstood something about UNIX?

Please take a look. It is easy to install, run, and understand.

**Comment by cheatfate** [comment]
Its not Unix but Linux problem, when you forking `epoll` descriptor with all registered handles inside of it duplicated across your forks, but `asyncdispatch/ioselectors` was initialized without this handles, so it cause problem, imho.

Currently i dont have enough time to help you with this case, sorry. But i'm pretty sure problem lies inside of inheritance of `epoll handle`.

**Comment by cheatfate** [comment]
Of course this also be a problem with this one issue (man epoll).

> Q6  Will closing a file descriptor cause it to be removed from all epoll sets automatically?

> A6  Yes, but be aware of the following point.  A file descriptor is a reference to an open file description (see open(2)).  Whenever a file descriptor is duplicated  via  dup(2),  dup2(2),  fcntl(2)
           F_DUPFD,  or  fork(2),  a new file descriptor referring to the same open file description is created.  An open file description continues to exist until all file descriptors referring to it have
           been closed.  A file descriptor is removed from an epoll set only after all the file descriptors referring to the underlying open file description  have  been  closed  (or  before  if  the  file
           descriptor  is  explicitly  removed using epoll_ctl(2) EPOLL_CTL_DEL).  This means that even after a file descriptor that is part of an epoll set has been closed, events may be reported for that
           file descriptor if other file descriptors referring to the same underlying file description remain open.

**Comment by cdunn2001** [comment]
Ok. That's helpful, actually.

What if I create all the pipes and forks up-front, and *then* register them all? I will try that....

(For the record, I'm doing it this way is because I had severe problems with memory when using a simple thread-pool. I have a Python/C program which uses Python `multiprocessing` module and runs faster, with more stable memory, than the Nim threadpool alternative. After some bug-fixes from Araq in Nim's memory manager, it seems stable on *most* platform, but on a particular VM at my company, virtual-memory explodes. I figured there would be fewer potential problems with simple forks.)

**Comment by cdunn2001** [comment]
Actually, this program already works that way. All forks/pipes are created, and *then* they are registered.

Note that there is no problem if we run only a single example. If you comment out `example1()`, then `example2()` will work fine.

Note also that all forks and fds are closed at the end of each example. You may be correct that the problem fork after an event-loop has been run, but the problem is that I am not allowed to clean-up the previous event-loop 100%. That is a bug in asyncdispath/selector/etc.

> Currently I dont have enough time to help you with this case, sorry.

I understand. But I still believe that this is a bug in the Nim async libraries, for an important use-case (i.e. the "re-use" use-case).

I don't have your level of expertise, but if I do find a way to get everything cleaned up and ready for new forks, would you be open to a pull-request? If so, should I be working with the `lib/upcoming/asyncdispatch.nim` version? When do you plan to switch to that?

**Comment by cdunn2001** [comment]
Here is something interesting. I added:
```nim
  proc setGlobalDispatcher*(disp: PDispatcher) =
    gDisp = disp
```
 And at the top of my examples, I added:
```nim
asyncdispatch.setGlobalDispatcher(asyncdispatch.newDispatcher())
```
Now, even the first example fails (on OSX; haven't tried Linux yet). There must be some global state beyond the global `PDispatcher`. That's the part the I need to be able to reset.

**Comment by cdunn2001** [comment]
Ah. It fails in that case because the global PDispatcher is created in the main thread and used in a fork. This is better:
```nim
asyncdispatch.setGlobalDispatcher(nil)
```
Then, the first `register()` in any new fork will use a brand new dispatcher.

With 2 examples, it still hangs in the 2nd, but we're definitely getting closer...

**Comment by cdunn2001** [comment]
Ok, I did have a bug. (A typo caused the wrong file-descriptors to be closed.) Otherwise, the basic theory is correct: It is important to release all state used by asyncdispatch before calling `posix.fork()` again. Closing.

(I've submitted PR  #5782 for `setGlobalDispatcher()` to make this possible, or safer.)

I'd also like to mention that `-d:nimDumpAsync` is *amazingly* helpful in debuging asyncmacro. Thanks for that!
